### PR TITLE
Ban replacements with unbacked SymInt on both sides

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124874
* #124864
* #123735
* #124785
* #124782
* #124739
* #124394
* __->__ #124316
* #124314
* #124310

Fixes https://github.com/pytorch/pytorch/issues/123854

Important comment:

```
                # Never replace unbacked symbols with other unbacked symbols.
                # This is error prone because you can cause references to
                # unbacked symbols to time travel backwards.  E.g.,
                #
                # u1 = x.item()
                # ... use of u1 ...
                # u2 = y.item()
                # u3 = z.item()
                # torch._check(u1 == u2 + u3)
                #
                # If you replace u1 with u2 + u3, then the use of u1 now
                # references u2 and u3 prior to them actually being bound at
                # runtime.  It's pretty inconvenient to setup control
                # dependencies for substitutions, so ban it entirely.
```

This is kind of risky for the internal MRS workstream, because we added these substitutions upon their request in the first place. Fortunately, we still allow substitutions to backed SymInts and constants, and I believe that is what is actually load bearing.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Comment by ezyang** [comment]
@ColinPeppler, I confirmed that this stack gets all the way to accuracy testing

```
AssertionError: Tensor-likes are not close!                                                                                                                            

Mismatched elements: 1 / 24 (4.2%)
Greatest absolute difference: 0.017822265625 at index (0,) (up to 0.01 allowed)
Greatest relative difference: 0.3671875 at index (0,) (up to 0.01 allowed)
```

on the repro you gave me at https://fb.workplace.com/groups/6829516587176185/posts/7079346225526552/

**Comment by ColinPeppler** [comment]
just curious, how are you able to test the changes in fbcode without a linked phabricator diff?

**Comment by ezyang** [comment]
I manually patched the diff in lol

**Comment by ezyang** [comment]
One thought here is that it might be OK to do a replacement, if we know the ordering of the unbacked symints. That is to say, it is OK to replace an older unbacked symint with an earlier one. Under the assumption that they don't get reordered, this replacement is guaranteed to be valid.

**Comment by ColinPeppler** [comment]
> it is OK to replace an older unbacked symint with an earlier one

re-order as an `u4` can be replaced with `u5` but not the other way around?

**Review by ColinPeppler — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -4240,31 +4244,32 @@ def _smart_symbol_sort(x):
...
+                # u2 = y.item()
+                # u3 = z.item()
+                # torch._check(u1 == u2 + u3)
+                #
+                # If you replace u1 with u2 + u3, then the use of u1 now
+                # references u2 and u3 prior to them actually being bound at
+                # runtime.  It's pretty inconvenient to setup control
+                # dependencies for substitutions, so ban it entirely.
+                if isinstance(lhs, sympy.Symbol) and free_unbacked_symbols(lhs) and not free_unbacked_symbols(rhs):
+                    # short-circuit when no solving is needed
```
Comment (ColinPeppler) [inline torch/fx/experimental/symbolic_shapes.py:4262]: looks like this was intended for the first 2 checks

**Comment by ezyang** [comment]
> re-order as an u4 can be replaced with u5 but not the other way around?

Yup. I'll do this in a follow up.

**Comment by lezcano** [comment]
Agreed about the replacement that is increasing in the given order (provided the order is stable under code transformations... which might be a bit of a big if...).
That would allow us to replace `u0` with `4*u1` if we happen to know that `u0 %  4 == 0`, which is nice.

File: test/test_dynamic_shapes.py [hunk 2]
```diff
@@ -562,14 +562,14 @@ def test_unbacked_substitution(self):
         _constrain_range_for_size(i0)
         _constrain_range_for_size(i1)
         self.assertTrue(expect_true(i0 == i1 * 4))
-        self.assertExpectedInline(str(i0), """4*u1""")
+        self.assertExpectedInline(str(i0), """u0""")
```
Comment (lezcano) [inline test/test_dynamic_shapes.py:565]: LOLOLOLOL

File: test/test_proxy_tensor.py [hunk 3]
```diff
@@ -1460,15 +1466,7 @@ def f(x1, x2, y):
...
-        r = str(make_fx(f, tracing_mode="symbolic")(torch.tensor(10), torch.tensor(10), torch.randn(10)).code).strip()
-        self.assertExpectedInline(r, """\
-def forward(self, x1_1, x2_1, y_1):
-    _local_scalar_dense = torch.ops.aten._local_scalar_dense.default(x1_1);  x1_1 = None
-    zeros = torch.ops.aten.zeros.default([_local_scalar_dense], device = device(type='cpu'), pin_memory = False);  _local_scalar_dense = None
-    _local_scalar_dense_1 = torch.ops.aten._local_scalar_dense.default(x2_1);  x2_1 = None
-    zeros_1 = torch.ops.aten.zeros.default([_local_scalar_dense_1], device = device(type='cpu'), pin_memory = False);  _local_scalar_dense_1 = None
-    add = torch.ops.aten.add.Tensor(y_1, 2);  y_1 = None
-    return add""")  # noqa: B950
+        make_fx(f, tracing_mode="symbolic")(torch.tensor(10), torch.tensor(10), torch.randn(10))
```
Comment (lezcano) [inline test/test_proxy_tensor.py:1469]: why this change?

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -2394,6 +2393,11 @@ def _recording(self):
         finally:
             self.is_recording = False
 
+    # Unlike set_replacement, this records a shapeenv event
+    @record_shapeenv_event()
+    def _rename_unbacked_to(self, orig_s: sympy.Expr, new_s: sympy.Expr):
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:2398]: both should be symbols, not Expr. Can you assert that they are also unbacked?

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -2394,6 +2393,11 @@ def _recording(self):
         finally:
             self.is_recording = False
 
+    # Unlike set_replacement, this records a shapeenv event
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:2396]: Remind me what are the implications of this other than clearing the caches?

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -4234,31 +4238,32 @@ def _smart_symbol_sort(x):
...
+                # u1 = x.item()
+                # ... use of u1 ...
+                # u2 = y.item()
+                # u3 = z.item()
+                # torch._check(u1 == u2 + u3)
+                #
+                # If you replace u1 with u2 + u3, then the use of u1 now
+                # references u2 and u3 prior to them actually being bound at
+                # runtime.  It's pretty inconvenient to setup control
+                # dependencies for substitutions, so ban it entirely.
```
Comment (lezcano) [inline torch/fx/experimental/symbolic_shapes.py:4254]: One way to do this is to declare that we just start using a substitution if all the unbacked symbols in it are defined. This also has the neat property that it "floats" the substitution as high up as they can go, meaning that
```
                # u1 = x.item()
                # ... use of u1 ...
                # u2 = y.item()
                # u3 = z.item()
                # .,.. use of u2, u3 ...
                # torch._check(u1 == u2 + u3)
```
becomes the example you wrote above

**Review by lezcano — COMMENTED** [review]
Seems fine, contingent on implementing https://github.com/pytorch/pytorch/pull/124316#issuecomment-2074460028.

File: test/test_proxy_tensor.py (hunk 3 above)
Comment (ezyang) [reply test/test_proxy_tensor.py:1469]: This test got xfailed. If I kept the expected graph, then the test might xfail even if it was passing and the expect string just needed updating.

File: torch/fx/experimental/symbolic_shapes.py (hunk 4 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:2398]: I'm going to add this assert in https://github.com/pytorch/pytorch/pull/124782 as I changed some details about how the call site works there

File: torch/fx/experimental/symbolic_shapes.py (hunk 5 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:2396]: If we don't record events, then translation validation won't replay this this event when doing bisection. Because rename unbacked to changes the state of ShapeEnv, it's important to track these.

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:4254]: While it might be possible to apply extra substitutions when we are evaluating guards, in general simplification can occur in unpredictable points whenever you access the .expr field of a SymNode. Also, applying extra substitutions when evaluating guards is also tricky, because we do multiple passes through the graph (e.g., when retracing) and whenever we start a new pass, we need to "reset" the set of unbacked symbols that are defined.

If we are relaxing this, I think I much prefer "unbacked SymInts have a topological order and you can only replace into earlier ordered SymInts." But I think this case is also difficult to do 100% correctly. Imagine you have:

```
y: f32[u0] = f(x)
z: f32[u1] = g(y)
```

We do a decomposition on g which introduces a fresh unbacked symbol:

```
y: f32[u0] = f(x)
z0: f32[u2] = g1(y)
z: f32[u1] = g2(z0)
```

Now the topological order of unbacked SymInts is not the same as a the order they were allocated. I need to know to insert u2 between u0 and u1 in the ordering, and I'm not even sure this maintenance is even easy to do in some of the optimization contexts.

This example is a bit contrived, but it's difficult enough that I don't really want to do anything about it unless we actually need it.

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:4254]: I guess I'll do something like, only unbacked SymInts that occurred on initial traversal of the graph have an order, and then anything that shows up later is banned from participating in replacements. Which I don't expect the latter case to ever occur.

File: torch/fx/experimental/symbolic_shapes.py (hunk 4 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:2398]: https://github.com/pytorch/pytorch/pull/124782/files#diff-03557db7303b8540f095b4f0d9cd2280e1f42f534f67d8695f756ec6c02d3ec7

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (lezcano) [reply torch/fx/experimental/symbolic_shapes.py:4254]: >we do multiple passes through the graph (e.g., when retracing) and whenever we start a new pass, we need to "reset" the set of unbacked symbols that are defined.

Sure, but we now have sources for symints, so the replacements could be done in terms of these sources, which should be pass-agnostic.

> nly unbacked SymInts that occurred on initial traversal of the graph have an order, and then anything that shows up later is banned from participating in replacements

this is also a reasonable approach

**Review by lezcano — APPROVED** [review]
Again, it would be good to allow some substitutions as per the discussion in https://github.com/pytorch/pytorch/pull/124316#discussion_r1577660433, but we can do that in another PR.

File: torch/fx/experimental/symbolic_shapes.py (hunk 6 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:4254]: You mean the keypaths. I don't see how this helps though. The keypath only makes sense relative to an FX node that produces the thing you run the keypath getter on, and you don't necessarily have an FX node when you're doing simplification.
